### PR TITLE
[5G: MVC][component: agw][type: enhancement] changes related to AMF-PR #4595

### PR DESCRIPTION
**Pull request by chandra-77**
Signed-off-by: chandra-77 <dangudubiyyam.c@altencalsoftlabs.com>

## Summary

Contains code for DNN IE, PDU address IE, DL nas transport and Pdu sessoion establishment accept. 
## Test Plan
S1AP integ_test are passed. Attached the log.
[s1ap_integ_test.log](https://github.com/magma/magma/files/6233989/s1ap_integ_test.log)

## Additional Information

- [ ] This change is backwards-breaking

### CODE REVIEW

File: lte/gateway/c/oai/tasks/nas5g/include/ies/M5GDNN.h [hunk 1]
```diff
@@ -0,0 +1,32 @@
+/*
+  Copyright 2020 The Magma Authors.
+  This source code is licensed under the BSD-style license found in the
+  LICENSE file in the root directory of this source tree.
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+*/
```
Comment (ssanadhya) [inline lte/gateway/c/oai/tasks/nas5g/include/ies/M5GDNN.h]: Use consistent header format across all c/c++ code

File: lte/gateway/c/oai/tasks/nas5g/include/ies/M5GDNN.h [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+*/
+#pragma once
+#include <sstream>
+#include <cstdint>
+
+using namespace std;
```
Comment (ssanadhya) [inline lte/gateway/c/oai/tasks/nas5g/include/ies/M5GDNN.h]: Not needed as you are explicitly specifying `std::string` in line 23

File: lte/gateway/c/oai/tasks/nas5g/src/M5GDLNASTransport.cpp [hunk 3]
```diff
@@ -143,6 +143,13 @@ int DLNASTransportMsg::EncodeDLNASTransportMsg(
     return encoded_result;
   else
     encoded += encoded_result;
+  if ((encoded_result =
+           dl_nas_transport->pdu_session_identity.EncodePDUSessionIdentityMsg(
+               &dl_nas_transport->pdu_session_identity, PDU_SESSION_IDENTITY,
+               buffer + encoded, len - encoded)) < 0)
+    return encoded_result;
+  else
+    encoded += encoded_result;
```
Comment (ssanadhya) [inline lte/gateway/c/oai/tasks/nas5g/src/M5GDLNASTransport.cpp]: Add braces around single statement conditional blocks

```suggestion
  if ((encoded_result =
           dl_nas_transport->pdu_session_identity.EncodePDUSessionIdentityMsg(
               &dl_nas_transport->pdu_session_identity, PDU_SESSION_IDENTITY,
               buffer + encoded, len - encoded)) < 0) {
    return encoded_result;
  } else {
    encoded += encoded_result;
    }
```

File: lte/gateway/c/oai/tasks/nas5g/src/M5GPDUSessionEstablishmentAccept.cpp [hunk 4]
```diff
@@ -89,6 +91,19 @@ int PDUSessionEstablishmentAcceptMsg::EncodePDUSessionEstablishmentAcceptMsg(
...
+               buffer + encoded, len - encoded)) < 0)
+    return encoded_result;
+  else
+    encoded += encoded_result;
+  if ((encoded_result = pdu_session_estab_accept->dnn.EncodeDNNMsg(
+           &pdu_session_estab_accept->dnn, DNN, buffer + encoded,
+           len - encoded)) < 0)
+    return encoded_result;
+  else
+    encoded += encoded_result;
```
Comment (ssanadhya) [inline lte/gateway/c/oai/tasks/nas5g/src/M5GPDUSessionEstablishmentAccept.cpp]: Add braces around single statement blocks too.

File: lte/gateway/c/oai/tasks/nas5g/src/ies/M5GDNN.cpp [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+   See the License for the specific language governing permissions and
+   limitations under the License.
+ */
+#include <sstream>
+#include <cstdint>
+#include <cstring>
+#include "M5GDNN.h"
+#include "M5GCommonDefs.h"
+
+using namespace std;
```
Comment (ssanadhya) [inline lte/gateway/c/oai/tasks/nas5g/src/ies/M5GDNN.cpp]: remove

File: lte/gateway/c/oai/tasks/nas5g/src/ies/M5GDNN.cpp [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+  // Checking IEI and pointer
+  CHECK_PDU_POINTER_AND_LENGTH_ENCODER(buffer, DNN_MIN_LENGTH, len);
+
+  if (iei > 0) {
+    CHECK_IEI_ENCODER(iei, (unsigned char) dnn_message->iei);
+    *buffer = iei;
+    encoded++;
+  }
+
+  MLOG(MDEBUG) << "EncodeDNN : ";
```
Comment (ssanadhya) [inline lte/gateway/c/oai/tasks/nas5g/src/ies/M5GDNN.cpp:54]: Use OAILOG

File: lte/gateway/c/oai/tasks/nas5g/include/ies/M5GAuthenticationParameterRAND.h [hunk 7]
```diff
@@ -20,6 +20,7 @@ namespace magma5g {
 class AuthenticationParameterRANDMsg {
  public:
 #define RAND_MIN_LEN 16
+#define RAND_MAX_LEN 16
```
Comment (ssanadhya) [inline lte/gateway/c/oai/tasks/nas5g/include/ies/M5GAuthenticationParameterRAND.h:23]: Is this being used in a future PR?

File: lte/gateway/c/oai/tasks/nas5g/include/ies/M5GAuthenticationParameterAUTN.h [hunk 8]
```diff
@@ -19,6 +19,7 @@ namespace magma5g {
 class AuthenticationParameterAUTNMsg {
  public:
 #define AUTN_MIN_LEN 16
+#define AUTN_MAX_LEN 16
```
Comment (ssanadhya) [inline lte/gateway/c/oai/tasks/nas5g/include/ies/M5GAuthenticationParameterAUTN.h:22]: Unused in this PR, is this being used in a future PR?

**Review by ssanadhya — CHANGES_REQUESTED** [review]

File: lte/gateway/c/oai/tasks/nas5g/include/ies/M5GDNN.h (hunk 1 above)
Comment (chandra-77) [reply lte/gateway/c/oai/tasks/nas5g/include/ies/M5GDNN.h]: Fixed the copyright for the DNN header and cpp files. we will fix for remaing all files in separate PR.

File: lte/gateway/c/oai/tasks/nas5g/include/ies/M5GDNN.h (hunk 2 above)
Comment (chandra-77) [reply lte/gateway/c/oai/tasks/nas5g/include/ies/M5GDNN.h]: fixed

File: lte/gateway/c/oai/tasks/nas5g/src/M5GDLNASTransport.cpp (hunk 3 above)
Comment (chandra-77) [reply lte/gateway/c/oai/tasks/nas5g/src/M5GDLNASTransport.cpp]: Added braces for this file, we will fix for remaing all files in separate PR

File: lte/gateway/c/oai/tasks/nas5g/src/M5GPDUSessionEstablishmentAccept.cpp (hunk 4 above)
Comment (chandra-77) [reply lte/gateway/c/oai/tasks/nas5g/src/M5GPDUSessionEstablishmentAccept.cpp]: Added

File: lte/gateway/c/oai/tasks/nas5g/src/ies/M5GDNN.cpp (hunk 5 above)
Comment (chandra-77) [reply lte/gateway/c/oai/tasks/nas5g/src/ies/M5GDNN.cpp]: removed

File: lte/gateway/c/oai/tasks/nas5g/src/ies/M5GDNN.cpp (hunk 6 above)
Comment (chandra-77) [reply lte/gateway/c/oai/tasks/nas5g/src/ies/M5GDNN.cpp:54]: We will change MLOG to OAILOG in separate PR.

File: lte/gateway/c/oai/tasks/nas5g/include/ies/M5GAuthenticationParameterRAND.h (hunk 7 above)
Comment (chandra-77) [reply lte/gateway/c/oai/tasks/nas5g/include/ies/M5GAuthenticationParameterRAND.h:23]: Its being used in amf

File: lte/gateway/c/oai/tasks/nas5g/include/ies/M5GAuthenticationParameterAUTN.h (hunk 8 above)
Comment (chandra-77) [reply lte/gateway/c/oai/tasks/nas5g/include/ies/M5GAuthenticationParameterAUTN.h:22]: Its being used in amf

**Review by ssanadhya — APPROVED** [review]

**Pull request merged**